### DISCUSSION TITLE
Is this a bug on MacOS? (SSL certificate verify failed, unable to get local issuer certificate)

### DISCUSSION THREAD

**Discussion by monocongo** [sqlalchemy/sqlalchemy#10297] [Usage Questions] [answered]
I have code that executes an SQL statement via `pandas.read_sql`, connecting to a SQL Server database:
```
   # build an SQLAlchemy Connectable that'll be used for database connections
    connection_string = f"mssql+pyodbc://{creds['username']}:{creds['password']}@{SERVER}:{PORT}/{DATABASE}?driver=ODBC Driver 18 for SQL Server"
    engine = create_engine(connection_string)

    # execute a simple extraction via pandas
    sql_query = "SELECT TOP 10 * FROM test_table;"
    df = pd.read_sql(sql=sql_query, con=engine, index_col='rowid')
```
This code works fine on a colleague's Intel-based Macbook but fails on my ARM-based Macbook (M2). This is the DBAPI error message I end up with:
```
sqlalchemy.exc.OperationalError: (pyodbc.OperationalError) ('08001', '[08001] [Microsoft][ODBC Driver 18 for SQL Server]SSL Provider: [error:0A000086:SSL routines::certificate verify failed:unable to get local issuer certificate] (-1) (SQLDriverConnect)')
```

If I hold my nose and swap out SQLAlchemy with pymssql then my problem goes away:
```
    # get a connection from pymssql
    connection = pymssql.connect(host=SERVER, port=PORT, user=creds['username'], password=creds['password'], database=DATABASE)

    # execute a simple extraction via pandas
    sql_query = "SELECT TOP 10 * FROM test_table;"
    df = pd.read_sql(sql=sql_query, con=engine, index_col='rowid')
```
This looks to me like an issue with the `pyodbc` library that's being called from within the SQL Server driver's code and the SQLAlchemy code just wraps that. Is this correct, and if so I need to ask the `pyodbc` developers about this instead, right?

My system environment details:
Macbook Pro
Apple M2 Pro chip
macOS Ventura 13.5.1

My conda environment was created like so:
```
conda create -n standardsqlalchemy python='3.10' pandas sqlalchemy
```
With the resulting package list:
```
% conda list
# packages in environment at /Users/jadams/miniconda3/envs/standardsqlalchemy:
#
# Name                    Version                   Build  Channel
blas                      1.0                    openblas
bottleneck                1.3.5           py310h96f19d2_0
bzip2                     1.0.8                h620ffc9_4
ca-certificates           2023.05.30           hca03da5_0
greenlet                  2.0.1           py310h313beb8_0
libcxx                    14.0.6               h848a8c0_0
libffi                    3.4.4                hca03da5_0
libgfortran               5.0.0           11_3_0_hca03da5_28
libgfortran5              11.3.0              h009349e_28
libopenblas               0.3.21               h269037a_0
llvm-openmp               14.0.6               hc6e5704_0
ncurses                   6.4                  h313beb8_0
numexpr                   2.8.4           py310hecc3335_1
numpy                     1.25.2          py310h3b2db8e_0
numpy-base                1.25.2          py310ha9811e2_0
openssl                   3.0.10               h1a28f6b_2
pandas                    2.0.3           py310h46d7db6_0
pip                       23.2.1          py310hca03da5_0
python                    3.10.12              hb885b13_0
python-dateutil           2.8.2              pyhd3eb1b0_0
python-tzdata             2023.3             pyhd3eb1b0_0
pytz                      2022.7          py310hca03da5_0
readline                  8.2                  h1a28f6b_0
setuptools                68.0.0          py310hca03da5_0
six                       1.16.0             pyhd3eb1b0_1
sqlalchemy                1.4.39          py310h1a28f6b_0
sqlite                    3.41.2               h80987f9_0
tk                        8.6.12               hb8d0fd4_0
tzdata                    2023c                h04d1e81_0
wheel                     0.38.4          py310hca03da5_0
xz                        5.4.2                h80987f9_0
zlib                      1.2.13               h5a0b063_0
```
What's interesting in the above is I don't see any ODBC Python packages in the virtual environment, so where is SQLAlchemy finding the `pyodbc` mentioned in the error message? Maybe it's somehow baked into the SQL Server driver code? There are so many layers to this onion...

Something else that may be at play is that there are multiple installations of `openssl` on my machine. I have both versions 1.1 and 3 installed via Homebrew, and we get another installation within the Python virtual environment (see package list above). All of these `openssl` installations contain certificates, so how can we make sure that the right one is being used? Because maybe that's the issue, it's confused as to which certificate needs to be in play and falls over as a result. Just a hunch.

Thanks for any suggestions or guidance.

**Comment by zzzeek** [comment, marked as answer]
im starting to get this also lately.  Add this to your URL

```
 connection_string = f"mssql+pyodbc://{creds['username']}:{creds['password']}@{SERVER}:{PORT}/{DATABASE}?driver=ODBC Driver 18 for SQL Server&TrustServerCertificate=yes"
  ````

**Reply by monocongo** [reply]
Hallelujah! Thank you for this help, @zzzeek I have been at this for days and this trick gets me over a big hump. Whether or not this is secure enough I don't know, seems like I remember seeing this mentioned as an insecure setting, but for now, it has me moving again.

**Reply by CaselIT** [reply]
> This looks to me like an issue with the pyodbc library that's being called from within the SQL Server driver's code and the SQLAlchemy code just wraps that. Is this correct, and if so I need to ask the pyodbc developers about this instead, right?

as you said, sqlalchemy just uses pyodbc, so ask them if there is something else to do to make it work on arm macs.

Also note that you can use pymssql also with sqlachemy, just use `mssql+pymssql://...`
